README.md
# Kafkatail

[![build](https://github.com/HandOfGod94/kafkatail/actions/workflows/build.yml/badge.svg)](https://github.com/HandOfGod94/kafkatail/actions/workflows/build.yml)
[![test](https://github.com/HandOfGod94/kafkatail/actions/workflows/test.yml/badge.svg)](https://github.com/HandOfGod94/kafkatail/actions/workflows/test.yml)

:construction: **Work In Progress** :construction:  
:construction: **Proceed with caution** :construction:

## Overview

CLI app to tail kafka logs on console from any topic having messages in any format: Plaintext, proto, ...

**Contents**  
- [Kafkatail](#kafkatail)
  - [Overview](#overview)
    - [Installation](#installation)
      - [MacOS using `homebrew`](#macos-using-homebrew)
      - [Linux using `snap`](#linux-using-snap)
      - [Windows using `scoop`](#windows-using-scoop)
      - [Go Toolchain](#go-toolchain)
      - [Github Releases](#github-releases)
    - [QuickStart](#quickstart)
    - [Usage](#usage)
    - [Examples](#examples)
    - [Known Limitation](#known-limitation)
    - [Development](#development)

### Installation

Feel free to use any of the available methods

#### MacOS using `homebrew`
```sh
brew install handofgod94/tap/kafkatail
```

#### Linux using `snap`
```sh
# currently only available on beta and edge channels
sudo snap install kafkatail --beta
sudo snap install kafkatail --edge
```

#### Windows using `scoop`
```cmd
scoop add handofgod94-bucket https://github.com/HandOfGod94/scoop-bucket.git
scoop install kafkatail
```

#### Go Toolchain
  + For Go 1.16 or higher
    ```sh
    go install github.com/handofgod94/kafkatail@v0.2.0
    ```

  + For go version <= 1.15
    ```sh
    GO111MODULE=on go get github.com/handofgod94/kafkatail@v0.2.0
    ```
    The go binary will be installed in `$GOPATH/bin`

#### Github Releases
* Download executable binaries from [Release](https://github.com/HandOfGod94/kafkatail/releases) page


### QuickStart

`$ kafkatail --version`
```
kafkatail version 0.2.0
```

### Usage

`$ kafkatail --help`
```
Tail kafka messages from any topic, of any wire format on console (plaintext, protobuf)

Usage:
  kafkatail [flags] topic

Flags:
  -b, --bootstrap_servers bootstrap_servers   list of kafka bootstrap_servers separated by comma
      --from_datetime string                  tail from specific past datetime in RFC3339 format (default "0001-01-01T00:00:00Z")
      --group_id group_id                     [Optional] kafka consumer group_id to be used for subscribing to topic
  -h, --help                                  help for kafkatail
      --include_paths include_paths           include_paths containing dependencies of proto. Required for `wire_format=proto`
      --message_type type                     proto message type to use for decoding . Required for `wire_format=proto`
      --offset int                            kafka offset to start consuming from. Possible Values: -1=latest, -2=earliest, n=nth offset (default -1)
      --partition either[int, string]         kafka partition to consume from. Can be int(0,1,2,...) or "all" (default all)
      --proto_file proto_file                 proto_file to be used for decoding kafka message. Required for `wire_format=proto`
  -v, --version                               version for kafkatail
      --wire_format wire_format[=plaintext]   Wire format of messages in topic (default plaintext)
```

### Examples
```sh
# tail messages from a topic
kafkatail --bootstrap_servers=localhost:9093 kafkatail-test

# tail proto messages from a topic
kafkatail --bootstrap_servers=localhost:9093 --wire_format=proto --proto_file=starwars.proto --include_paths="../testdata" --message_type=Human kafkatail-test-proto

# tail messages from an offset. Default: -1 (latest). For earliest, use offset=-2
kafkatail --bootstrap_servers=localhost:9093 --offset=12 kafkatail-test-base

# tail messages from specific time
kafkatail --bootstrap_servers=localhost:9093 --from_datetime=2021-06-28T15:04:23Z kafkatail-test-base

# tail messages from specific partition. Default: 0
kafkatail --bootstrap_servers=localhost:9093 --partition=5 kafkatail-test-base

# tail from multiple partitions, using group_id
kafkatail --bootstrap_servers=localhost:9093 --group_id=myfoo kafka-consume-gorup-id-int-test

# tail from multiple partitions WITHOUT group_id
kafkatail --bootstrap_servers=localhost:9093 --partition=all kafka-consume-gorup-id-int-test
# or simply
kafkatail --bootstrap_servers=localhost:9093 kafka-consume-gorup-id-int-test
	
```

### Known Limitation

* `--offset` option doesn't work with `--gorup_id`
* No `avro`, `thirft`, `<custom>` decoding support yet.  
  Only supports `protobuf` and `plaintext` for now

### Development
* Minimum go version: `1.18`
* Ruby version: `2.7.1` (for running cucumber-tests)

```sh
# to run dev checks and create binary
make

# to run test
make test

# to run integration tests
# it requires ruby to be installed
# download cucumber dependencies
make cucumber-setup

# run cucumber tests
make cucumber-test
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html),
and is generated by [Changie](https://github.com/miniscruff/changie).

## v0.2.0 - 2022-06-13

### Added
* ability to tail from multiple partitions without setting group_id, using `--partition=all` flag

### Changed
* Go version upgraded from 1.16 to 1.18

## v0.1.5 - 2021-12-31

### Added
* Snap release for linux

## v0.1.4 - 2021-12-30

### Added
* Windows Scoop Release

## v0.1.3 - 2021-10-29

### Added
* Homebrew release
* Validations for proto specific flags

## v0.1.2 - 2021-08-08

### Added
* close KafkaReader on app termination (#1)
* Examples for kafkatail help command

### Fixed
* consume from all partition when group_id is specified (#2)

## v0.1.1 - 2021-07-07

### Fixed
* Validation error when whitespace is used instead of `=` in command flags

## v0.1.0 - 2021-07-03

### Added
* Tail from specific datetime for a topic
* Tail from specific partition for a topic
* Tail from a specific offset on a topic partition
* Tail proto messages after decoding from a topic
* version info flag
* Tail plaintext messages from a topic


cmd/root.go
package cmd

import (
	"context"
	"fmt"
	"log"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/handofgod94/kafkatail/consumer"
	"github.com/handofgod94/kafkatail/wire"
	"github.com/samber/mo"
	"github.com/spf13/cobra"
	"github.com/thediveo/enumflag"
)

var (
	bootstrapServers   []string
	groupID            string
	wireForamt         wire.Format
	offset             int64
	fromDateTime       string
	parsedFromDateTime time.Time
	protoFile          string
	includePaths       []string
	messageType        string
	partition          = PartitionFlag{value: mo.Right[int]("all")}
)

const appVersion = "0.2.0"

type status = int

// rootCmd represents the base command when called without any subcommands
var rootCmd = &cobra.Command{
	Use:     "kafkatail [flags] topic",
	Short:   "Tail kafka logs of any wire format",
	Long:    `Tail kafka messages from any topic, of any wire format on console (plaintext, protobuf)`,
	Version: appVersion,
	Args:    cobra.MinimumNArgs(1),
	Example: `
	# tail messages from a topic
	kafkatail --bootstrap_servers=localhost:9093 kafkatail-test

	# tail proto messages from a topic
	kafkatail --bootstrap_servers=localhost:9093 --wire_format=proto --proto_file=starwars.proto --include_paths="../testdata" --message_type=Human kafkatail-test-proto

	# tail messages from an offset. Default: -1 (latest). For earliest, use offset=-2
	kafkatail --bootstrap_servers=localhost:9093 --offset=12 kafkatail-test-base

	# tail messages from specific time
	kafkatail --bootstrap_servers=localhost:9093 --from_datetime=2021-06-28T15:04:23Z kafkatail-test-base

	# tail messages from specific partition. Default: 0
	kafkatail --bootstrap_servers=localhost:9093 --partition=5 kafkatail-test-base

	# tail from multiple partitions, using group_id
	kafkatail --bootstrap_servers=localhost:9093 --group_id=myfoo kafka-consume-gorup-id-int-test

	# tail from multiple partitions WITHOUT group_id
	kafkatail --bootstrap_servers=localhost:9093 --partition=all kafka-consume-gorup-id-int-test
	# or simply
	kafkatail --bootstrap_servers=localhost:9093 kafka-consume-gorup-id-int-test
	`,
	PreRunE: func(cmd *cobra.Command, args []string) (err error) {
		formatFlag := cmd.Flags().Lookup("wire_format")
		if formatFlag.Value.String() == "proto" {
			cmd.MarkFlagRequired("proto_file")
			cmd.MarkFlagRequired("include_paths")
			cmd.MarkFlagRequired("message_type")
		}

		parsedFromDateTime, err = time.Parse(time.RFC3339, fromDateTime)
		if err != nil {
			return err
		}

		return nil
	},
	RunE: func(cmd *cobra.Command, args []string) error {
		topic := args[0]
		cf := consumer.NewFactory(
			bootstrapServers,
			groupID,
			topic,
			partition.Get(),
			offset,
			parsedFromDateTime,
			wireForamt,
			protoFile,
			messageType,
			includePaths,
		)

		konsumer, err := cf.CreateConsumer()
		if err != nil {
			return err
		}

		decoder, err := cf.CreateDecoder()
		if err != nil {
			return err
		}
		resultChan := konsumer.Consume(context.Background(), decoder)
		exitCode := <-receiveMessages(resultChan)

		konsumer.Close()
		log.Printf("stopping application, with exitcode: %d", exitCode)
		os.Exit(exitCode)

		return nil
	},
}

func Execute() {
	cobra.CheckErr(rootCmd.Execute())
}

func init() {
	zeroTime := time.Date(1, 1, 1, 0, 0, 0, 0, time.UTC).Format(time.RFC3339)

	rootCmd.Flags().StringSliceVarP(&bootstrapServers, "bootstrap_servers", "b", []string{}, "list of kafka `bootstrap_servers` separated by comma")
	rootCmd.Flags().StringVar(&groupID, "group_id", "", "[Optional] kafka consumer `group_id` to be used for subscribing to topic")
	rootCmd.Flags().Int64Var(&offset, "offset", -1, "kafka offset to start consuming from. Possible Values: -1=latest, -2=earliest, n=nth offset")
	rootCmd.Flags().Var(&partition, "partition", `kafka partition to consume from. Can be int(0,1,2,...) or "all"`)
	rootCmd.Flags().StringVar(&fromDateTime, "from_datetime", zeroTime, "tail from specific past datetime in RFC3339 format")
	rootCmd.Flags().StringVar(&messageType, "message_type", "", "proto message `type` to use for decoding . Required for `wire_format=proto`")
	rootCmd.Flags().StringSliceVar(&includePaths, "include_paths", []string{}, "`include_paths` containing dependencies of proto. Required for `wire_format=proto`")
	rootCmd.Flags().StringVar(&protoFile, "proto_file", "", "`proto_file` to be used for decoding kafka message. Required for `wire_format=proto`")
	rootCmd.Flags().Var(enumflag.New(&wireForamt, "wire_format", wire.FormatIDs, enumflag.EnumCaseSensitive),
		"wire_format",
		"Wire format of messages in topic",
	)

	rootCmd.Flags().Lookup("wire_format").NoOptDefVal = "plaintext"

	rootCmd.MarkFlagRequired("bootstrap_servers")
}

func receiveMessages(resultChan <-chan consumer.Result) <-chan status {
	sigs := make(chan os.Signal, 1)
	exitCode := make(chan status)

	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)

	go func() {
		loop := true
		for loop {
			select {
			case result := <-resultChan:
				if result.Err != nil {
					loop = false
					log.Print("error while consuming messages:", result.Err)
					exitCode <- 1
				}
				fmt.Println(result.Message)
			case <-sigs:
				loop = false
				exitCode <- 0
			}
		}
	}()

	return exitCode
}
